dali/operators/signal/fft/spectrogram.cc

```diff
@@ -0,0 +1,265 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <vector>
+#include "dali/operators/audio/fft/spectrogram.h"
+#include "dali/kernels/kernel_manager.h"
+#include "dali/kernels/signal/window/extract_windows_cpu.h"
+#include "dali/kernels/signal/fft/fft_cpu.h"
+#include "dali/core/static_switch.h"
+#include "dali/pipeline/data/views.h"
+
+#define SPECTROGRAM_SUPPORTED_NDIMS (1, 2)
+
+namespace dali {
+
+DALI_SCHEMA(Spectrogram)
+  .DocStr(R"code(Produces a spectrogram from a 1D signal (e.g. audio). Input data is expected
+to be single channel (1D shape `(time)`) or multi channel in planar layout (channel, time) 32 bit
+float tensor)code")
+  .NumInput(1)
+  .NumOutput(1)
+  .AddOptionalArg("nfft",
+    R"code(Size of the FFT. The number of bins created in the output is `nfft // 2 + 1`
+(positive part of the spectrum only).)code",
+    -1)
+  .AddOptionalArg("window_length",
+    R"code(Window size (in number of samples))code",
+    512)
+  .AddOptionalArg("window_step",
+    R"code(Step betweeen the STFT windows)code",
+    256)
+  .AddOptionalArg("window_fn",
```

Comment:
Why not AdditionalInput?

Reply:
My guess is that it's not likely to be per-sample.

Reply:
It's a small list of float32 numbers (nfft size) describing the function. The user will just bake it in python (or use scipy or similar to generate the function) and pass it to the operator as a python list.

Reply:
I was asking the thing that @mzient just answered. Fair.

Reply:
I think we should have some utilities for generating common window functions - even if just in Python.